python/foxglove-sdk/src/websocket_server.rs

```diff
@@ -0,0 +1,193 @@
+use crate::errors::PyFoxgloveError;
+use foxglove::{
+    websocket::{Client, ClientChannelView, ServerListener},
+    WebSocketServer, WebSocketServerBlockingHandle,
+};
+use pyo3::{
+    prelude::*,
+    types::{PyBytes, PyDict},
+};
+use std::sync::Arc;
+use std::time;
+
+/// A mechanism to register callbacks for handling client message events.
+#[pyclass(name = "ServerListener", module = "foxglove")]
+pub struct PyServerListener {
+    listener: Py<PyAny>,
+}
+
+#[pymethods]
+impl PyServerListener {
+    #[new]
+    fn new(listener: Py<PyAny>) -> Self {
+        PyServerListener { listener }
+    }
+}
+
+/// A client connected to a running websocket server.
+#[pyclass(name = "Client", module = "foxglove")]
+pub struct PyClient {
+    #[pyo3(get)]
+    id: u32,
+}
+
+/// Information about a client channel.
+#[pyclass(name = "ClientChannelView", module = "foxglove")]
+pub struct PyClientChannelView {
+    #[pyo3(get)]
+    id: u32,
+    #[pyo3(get)]
+    topic: String,
+}
+
+/// Implementations of ServerListener which call the python methods. foxglove/__init__.py defines
+/// the `ServerListener` protocol for callers, since a `pyclass` cannot extend Python classes:
+/// https://github.com/PyO3/pyo3/issues/991
+///
+/// The ServerListener protocol implements all methods as no-ops by default; users extend this with
+/// desired functionality.
+///
+/// Methods on the listener interface do not return Results; any errors are logged, assuming the
+/// user has enabled logging.
+impl ServerListener for PyServerListener {
+    fn on_message_data(&self, client: Client, channel: ClientChannelView, payload: &[u8]) {
+        let client_info = PyClient {
+            id: client.id().into(),
+        };
+
+        let channel_view = PyClientChannelView {
+            id: channel.id().into(),
+            topic: channel.topic().to_string(),
+        };
+
+        let result: PyResult<()> = Python::with_gil(|py| {
+            let kwargs = PyDict::new(py);
```

Comment:
We're creating this temporary dict here to call every callback. Why not use positional arguments?

The user can still choose to use kwargs if they want, but then at least they're aware of the overhead they're adding.

Reply:
👍 updated